stl/inc/xlocnum

```diff
@@ -750,366 +746,315 @@ private:
-
-            if (_Px == _Leading) { // change "999..." to "1000..." and scale _Pten
-                *_Px = '1';
-                ++_Pten;
+        if (_Has_unaccumulated_digits) { // increment last digit in memory of those lost
+            char& _Last_got_digit = _Ptr[-1] == _Decimal_point ? _Ptr[-2] : _Ptr[-1];
+            if (_Last_got_digit == '0' || _Last_got_digit == (_Parse_hex ? '8' : '5')) {
+                ++_Last_got_digit;
             }
         }
 
+        const _Elem _Lower_exp_wc = _Atoms[_Parse_hex ? _Offset_lower_p : _Offset_lower_e]; // 'e' for dec, 'p' for hex
+        const _Elem _Upper_exp_wc = _Atoms[_Parse_hex ? _Offset_upper_p : _Offset_upper_e]; // 'E' for dec, 'P' for hex
+
+        bool _Exponent_part_negative = false;
+        ptrdiff_t _Exponent_part     = 0;
         if (_Seendigit && _First != _Last
-            && (*_First == _Atoms[_Numget_eoff + 1]
-                || *_First == _Atoms[_Numget_eoff])) { // 'e' or 'E', collect exponent
-            *_Ptr++ = 'e';
+            && (*_First == _Lower_exp_wc || *_First == _Upper_exp_wc)) { // collect exponent
             ++_First;
             _Seendigit   = false;
             _Significant = 0;
 
             if (_First != _Last) {
-                if (*_First == _Atoms[_Numget_signoff + 1]) { // gather plus sign
-                    *_Ptr++ = '+';
+                if (*_First == _Positive_sign) { // gather plus sign
                     ++_First;
-                } else if (*_First == _Atoms[_Numget_signoff]) { // gather minus sign
-                    *_Ptr++ = '-';
+                } else if (*_First == _Negative_sign) { // gather minus sign
+                    _Exponent_part_negative = true;
                     ++_First;
                 }
             }
 
-            for (; _First != _Last && *_First == _Atoms[0]; ++_First) { // strip leading zeros
+            for (; _First != _Last && *_First == _Zero_wc; ++_First) { // strip leading zeros
                 _Seendigit = true;
             }
 
             if (_Seendigit) {
                 *_Ptr++ = '0'; // put one back
             }
 
-            for (; _First != _Last && (_Idx = _Find_elem(_Atoms, *_First)) < 10; _Seendigit = true, (void) ++_First) {
-                if (_Significant < _MAX_EXP_DIG) { // save a significant exponent digit
-                    *_Ptr++ = _Src[_Idx];
-                    ++_Significant;
+            for (size_t _Idx; _First != _Last && (_Idx = _STD _Find_elem(_Atoms, *_First)) < _Offset_dec_digit_end;
+                 _Seendigit = true, (void) ++_First) {
+                if (_Exponent_part < PTRDIFF_MAX / 10
+                    || (_Exponent_part == PTRDIFF_MAX / 10
+                        && static_cast<ptrdiff_t>(_Idx) <= PTRDIFF_MAX % 10)) { // save a significant exponent digit
+                    _Exponent_part = _Exponent_part * 10 + static_cast<ptrdiff_t>(_Idx);
+                } else {
+                    _Exponent_part = PTRDIFF_MAX; // saturated

```

Comment:
The saturation may be incorrect for certain extremely lengthy input and deliciated exponent part. E.g. the input may be a hexadecimal representation whose length is greater than `PTRDIFF_MAX / 4`, while its exponent part is so precisely specified (which may need to be greater than `PTRDIFF_MAX`) that the result is a non-zero finite number.